### DISCUSSION TITLE
Requesting clarification on number of threads and MPI+OpenMP

### DISCUSSION THREAD

**Discussion by bruno-rescale** [OpenRadioss/OpenRadioss#556] [Q&A] [answered]
Hi,

I have a 2-node cluster, 36 cores/node (72 total cores). I want to fully subscribe all the available cores and I submitted the following command (ignore blank input file):

 export OMP_NUM_THREADS=36
  ./starter_linux64_gf -i [Starter input file] -np 2
  mpiexec -n 72 ./engine_linux64_gf_ompi -i [Engine input file]

The idea is that it will launch 2 SPMD domains (1 SMPD in each node) and in each SPMD, it will run on 36 threads therefore 36 cores. However, I got the following error:

``` ROOT: BIRD_WINDSHIELD_v1  RESTART: 0001
 THE REQUIRED NUMBER OF MPI PROCESSES DOES NOT MATCH MPIRUN
 PLEASE, RUN WITH THE PROPER NUMBER OF MPI PROCESSES

  E R R O R     T E R M I N A T I O N
  TOTAL NUMBER OF CYCLES  :           0
```

If the number of MPI processes have to match, it appears to say the number of SPMD domains (-np) has to match the number of MPI processes (mpiexec -n). If I want to fuly subscribe, the only way it lets me submit is if "-np" and "mpiexec -n" are 72 like this:

 export OMP_NUM_THREADS=1
  ./starter_linux64_gf -i [Starter input file] -np 72
  mpiexec -n 72 ./engine_linux64_gf_ompi -i [Engine input file]

The above actually works and solved the "bird windshield" example but this forces each MPI process to be of 1 thread / 1 core. Is this a limitation? How can I do hybrid? Say I want to do 2 SPMD domains per node, each MPI process containing 36 cores/threads.

**Comment by elequiniou** [comment, marked as answer]
Hi Bruno, 

Thank you very much for your question and interest in OpenRadioss.
The way to run hybrid is documented under [How to Run OpenRadioss](https://github.com/OpenRadioss/OpenRadioss/blob/main/INSTALL.md)

The -np of starter determines the number of MPI domains. It needs to match the -n of the mpiexec command.
For low number of cores/nodes, I would recommend to run with "MPI only" as you did with -np 72.
For large number of nodes, running few threads per MPI can help, say 2 or 4 threads per MPI should be good.

Example to run with 2 threads per MPI in your case:

export OMP_NUM_THREADS=2
./starter_linux64_gf -i [Starter input file] -np 36
mpiexec -n 36 ./engine_linux64_gf_ompi -i [Engine input file]

I hope this clarifies your question and I let you experiment with your specific configuration.

**Comment by bruno-rescale** [comment]
Thank you very much for the thorough feedback @elequiniou . This makes perfect sense now. I'm able to setup OpenRadioss job runs now.
